File: website/backend/analytics.py
"""Receita no Google Analytics, pelo lado do servidor.

POR QUE NÃO NO NAVEGADOR. O caminho óbvio seria disparar `gtag('event',
'purchase')` na página de retorno do checkout. Ele perde dinheiro de duas
formas, e as duas são enviesadas:

  1. O MercadoPago leva o usuário pra fora do site. Quem paga por PIX
     costuma fechar a aba na tela de confirmação e nunca volta pro
     /checkout/sucesso -- o pagamento acontece, o evento não.
  2. Público mobile e nicho de aposta tem taxa alta de bloqueador. O que o
     bloqueador derruba não é uma amostra aleatória da base.

Somando os dois, o relatório mostraria menos receita do que existe, e menos
justamente nos canais que mais convertem. Pior que não medir.

Aqui o evento sai de `_apply_approved_payment`, que é o único ponto onde um
pagamento vira VIP (webhook, retorno do checkout e os dois botões do admin
passam todos por lá) e que já é idempotente por `mp_payment_id`. Um pagamento,
um evento, independente de o usuário ter voltado pro site.

O PREÇO DISSO é o `client_id`: sem ele o GA registra a receita como sessão
nova e direta, e a pergunta que justifica o trabalho todo -- de qual canal veio
quem paga -- fica sem resposta. Por isso o checkout captura o `_ga` do
navegador e guarda em `users.ga_client_id`; este módulo só o lê de volta.
"""
import logging
import os

import requests

from runtime_env import is_production

logger = logging.getLogger(__name__)

GA_MEASUREMENT_ID = os.getenv("GA_MEASUREMENT_ID", "")
GA_API_SECRET     = os.getenv("GA_API_SECRET", "")
GA_ENDPOINT       = "https://www.google-analytics.com/mp/collect"


def parse_ga_cookie(raw: str) -> str:
    """Extrai o client_id do cookie `_ga`.

    O formato é `GA1.1.<client_id>`, e o client_id em si tem um ponto no meio
    (`1234567890.1699999999`) -- então o corte é pelos dois primeiros campos,
    não por split('.') simples, que devolveria só metade do id.
    """
    parts = (raw or "").strip().split(".")
    if len(parts) < 4:
        return ""
    return f"{parts[2]}.{parts[3]}"


def send_purchase(client_id: str, user_id: int, payment_id: str,
                  plan_key: str, plan_title: str, amount: float) -> None:
    """Manda um `purchase` pro GA4 via Measurement Protocol.

    Nunca levanta exceção: é chamado depois do commit que ativa o VIP, e
    analytics não pode derrubar (nem parecer que derrubou) um pagamento que
    já entrou.
    """
    if not (GA_MEASUREMENT_ID and GA_API_SECRET):
        return
    # Só produção. O noprod aponta pro banco de produção e um teste de
    # pagamento lá dentro contaminaria o relatório real com receita que não
    # existe · mesmo motivo pelo qual `is_production` existe pra cota de API.
    if not is_production():
        logger.info("[GA] Fora de produção · purchase de %s não enviado.", payment_id)
        return
    if not client_id:
        # Sem client_id o evento entraria como sessão nova e direta, inflando
        # "Direct" e roubando o crédito do canal que de fato trouxe a venda.
        # Registrar receita errada é pior do que não registrar.
        logger.info("[GA] Sem client_id pro user %s · purchase de %s não enviado.", user_id, payment_id)
        return

    payload = {
        "client_id": client_id,
        # user_id interno, nunca e-mail ou CPF: mandar PII pro GA viola os
        # termos de uso e pode derrubar a propriedade inteira.
        "user_id": str(user_id),
        "non_personalized_ads": True,
        "events": [{
            "name": "purchase",
            "params": {
                # transaction_id é o que faz o GA descartar duplicata se este
                # pagamento for reprocessado por outro caminho.
                "transaction_id": str(payment_id),
                "value":          round(float(amount), 2),
                "currency":       "BRL",
                "items": [{
                    "item_id":       plan_key,
                    "item_name":     plan_title,
                    "item_category": "assinatura",
                    "price":         round(float(amount), 2),
                    "quantity":      1,
                }],
            },
        }],
    }

    try:
        resp = requests.post(
            GA_ENDPOINT,
            params={"measurement_id": GA_MEASUREMENT_ID, "api_secret": GA_API_SECRET},
            json=payload,
            timeout=5,
        )
        # O Measurement Protocol responde 204 pra praticamente tudo, inclusive
        # payload inválido. Erro de formato só aparece no endpoint /debug, então
        # 2xx aqui significa "chegou", não "está correto".
        if resp.status_code >= 300:
            logger.warning("[GA] purchase %s recusado: HTTP %s", payment_id, resp.status_code)
        else:
            logger.info("[GA] purchase %s enviado · R$ %.2f", payment_id, amount)
    except Exception as e:
        logger.warning("[GA] Falha ao enviar purchase %s: %s", payment_id, e)

File: website/frontend/src/lib/analytics.ts
/**
 * Eventos de funil no GA4.
 *
 * O QUE ESTE ARQUIVO **NÃO** FAZ: `purchase`. A receita é enviada pelo
 * servidor, de `_apply_approved_payment` (ver backend/analytics.py), e o
 * motivo está lá: quem paga por PIX fecha a aba na tela do MercadoPago e
 * nunca volta pro site, e bloqueador de anúncio é comum neste público. Medir
 * a venda daqui perderia dinheiro justamente nos canais que mais convertem.
 *
 * O que sobra pro navegador é o funil ANTES da venda · ver a página de
 * planos, iniciar o checkout, criar conta, entrar. Esses eventos só existem
 * do lado do cliente: o servidor não sabe que alguém olhou a tabela de
 * preços. São eles que preenchem a "Jornada de compra" e a atribuição de
 * canal no relatório.
 *
 * NADA DE DADO PESSOAL AQUI. Nome, e-mail, telefone e CPF não podem ir pro
 * GA · além de violar os termos de uso, dá pra derrubar a propriedade
 * inteira. O identificador de usuário é o id interno, e ele é enviado pelo
 * servidor, não daqui.
 */

const MOEDA = 'BRL'

type Parametros = Record<string, unknown>

/**
 * O mínimo de um plano que o GA precisa. Aceita o `Plan` de usePlans sem
 * conversão · o preço continua vindo de /api/payments/plans, que é a mesma
 * tabela que o MercadoPago cobra.
 */
export interface PlanoGA {
  id: string
  title?: string
  price: number
}

interface ItemGA {
  item_id: string
  item_name: string
  item_category: string
  price: number
  quantity: number
}

/**
 * Dispara um evento se o gtag estiver de pé.
 *
 * Silencioso de propósito quando não está: o script vem de fora e cai por
 * bloqueador com frequência neste público. Medição nunca pode derrubar, nem
 * atrasar, o fluxo de quem está comprando.
 */
function evento(nome: string, parametros: Parametros = {}): void {
  try {
    const gtag = (window as unknown as { gtag?: (...args: unknown[]) => void }).gtag
    if (typeof gtag !== 'function') return
    gtag('event', nome, parametros)
  } catch {
    /* medição não quebra tela */
  }
}

function paraItem(plano: PlanoGA): ItemGA {
  return {
    item_id: plano.id,
    item_name: plano.title || plano.id,
    item_category: 'assinatura',
    price: Number(plano.price) || 0,
    quantity: 1,
  }
}

/** Alguém abriu a tabela de preços. */
export function viuOsPlanos(planos: Array<PlanoGA>): void {
  if (!planos.length) return
  evento('view_item_list', {
    item_list_id: 'planos',
    item_list_name: 'Planos de assinatura',
    items: planos.map(paraItem),
  })
}

/** Alguém escolheu um plano e foi pro checkout. */
export function escolheuPlano(plano: PlanoGA): void {
  evento('select_item', { item_list_id: 'planos', items: [paraItem(plano)] })
}

/**
 * Último passo medido no navegador: o clique que cria o pagamento.
 *
 * Depois daqui o usuário sai do site pro MercadoPago, e o que acontece lá
 * só o servidor fica sabendo.
 */
export function iniciouCheckout(plano: PlanoGA): void {
  evento('begin_checkout', {
    currency: MOEDA,
    value: Number(plano.price) || 0,
    items: [paraItem(plano)],
  })
}

export function criouConta(): void {
  evento('sign_up', { method: 'email' })
}

export function entrou(): void {
  evento('login', { method: 'email' })
}
